### PR TITLE
[IA-2812] Unique Galaxy disk per workspace

### PR DESCRIPTION
**Pull request by gpcarr**
[Bug](https://broadworkbench.atlassian.net/browse/IA-2812): If a user has workspaces that were created before the PPW migration, it is possible for that user to create galaxy instances in different workspaces but the same google project. [Leonardo does not allow the same disk to exist in 2 different workspaces](https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala#L167). so the UI needs to create a new disk per workspace (if they are in the same project pre PPW migration). 

This PR fixes the bug mentioned. Tested by using two workspaces that were created pre PPW migration: `callisto-dev/gab-test` and `callisto-dev/gab-test copy`.  Created a galaxy in `gab-test`, deleted that galaxy but kept the disk. Then created a galaxy in `gab-test copy` and validated that the UI does not attach the disk from `gab-test` and creates a new disk and validated that Leo no longer throws an error and galaxy created successfully.

<!--
Hello, friend!
Remember to mention what sort of testing/verification you performed in the course of building this PR, i.e. checking functionality in the browser locally.
Thanks!
--!>

### CODE REVIEW

**Comment by kyuksel** [comment]
@gpcarr Before setting out to review and test the PR, I thought I’d reproduce the error on `dev` first and to my surprise, I was able to have two `Galaxy`s running concurrently in different workspaces in the same project without any errors 🤔

**Comment by kyuksel** [comment]
@gpcarr Nevermind - got clarification that the error is seen only with pre-PPW workspaces, which makes sense.

**Comment by kyuksel** [comment]
With this PR's changes, I see that I can now create two `Galaxy`s in the same project, using distinct workspaces. I've noticed two issues:

1. The Cloud Environments page still displays the warning triangles for having multiple cloud envs/disks.

image

2. When I had the first Galaxy app running and was going to create the second Galaxy app, the Galaxy modal seemed to have selected the existing disk as shown below although a new disk ended up being created. Perhaps there's still something to sort out with the logic that detects an existing disk?

image

For testing, I used Rob's pre-PPW workspaces, `test-rstudio` and `rstudio test`, and the `callisto-dev` project.

File: src/components/GalaxyModal.js
Comment (kyuksel) [inline src/components/GalaxyModal.js]: Given the deconstruction

```javascript
workspace: { workspace: { namespace, bucketName, name: workspaceName, googleProject } }
```

in `GalaxyModalBase` input definition

```javascript
export const GalaxyModalBase = Utils.withDisplayName('GalaxyModal')(
  ({
    onDismiss, onSuccess, apps, galaxyDataDisks, workspace, workspace: { workspace: { namespace, bucketName, name: workspaceName, googleProject } },
    isAnalysisMode = false
  }) => {…}
```

don't we already have `workspace.workspace.name`?

```suggestion
    const currentDataDisk = currentPersistentDisk(apps, galaxyDataDisks, workspaceName)
```

File: src/pages/Environments.js
Comment (kyuksel) [inline src/pages/Environments.js]: Could you remind me why we still need to make the second `Disks.list` call instead of filtering through the results of the previous `Disks.list` call?

**Comment by gpcarr** [comment]
> With this PR's changes, I see that I can now create two `Galaxy`s in the same project, using distinct workspaces. I've noticed two issues:
> 
> 1. The Cloud Environments page still displays the warning triangles for having multiple cloud envs/disks.
> 
> image
> 
> 1. When I had the first Galaxy app running and was going to create the second Galaxy app, the Galaxy modal seemed to have selected the existing disk as shown below although a new disk ended up being created. Perhaps there's still something to sort out with the logic that detects an existing disk?
> 
> image
> 
> For testing, I used Rob's pre-PPW workspaces, `test-rstudio` and `rstudio test`, and the `callisto-dev` project.

Taking a look at both of these issues now and will update the PR accordingly!

**Comment by kyuksel** [comment]
@gpcarr FYI for when you get back to this PR:

@cahrens's [recent PRs](https://github.com/DataBiosphere/terra-ui/pulls?q=is%3Apr+draft%3Afalse+author%3Acahrens+) have likely touched pieces you'll be updating so you may want to rebase before making any changes 😊 

image

File: src/pages/Environments.js
Comment (cahrens) [reply src/pages/Environments.js]: With my PR, we will just be making a single Disk.list call. See https://github.com/DataBiosphere/terra-ui/pull/2702/files#diff-db07f65dd0ba5edbd535b2a231709c1f5901f4e1c43f39c333165d97c1abacc4R133

File: src/pages/workspaces/workspace/WorkspaceContainer.js
Comment (cahrens) [inline src/pages/workspaces/workspace/WorkspaceContainer.js:183]: Here also there will be  single call. I don't think the workspace name will ever be used though by WorkspaceContainer… why are you including the label here?

Comment (cahrens) [inline src/pages/workspaces/workspace/WorkspaceContainer.js]: Here also this will be just a single call after my PR merges.

File: src/pages/Environments.js
Comment (cahrens) [reply src/pages/Environments.js]: I don't think though that the Environments page needs the workspace name. It does not call `currentPersistentDisk`. The Environments page already partitions the disks by Google project, using disksByProject (the project property of the disk).

In general, why use saturnWorkspaceName instead of googleProject? With project per workspace, shouldn't they be the same thing?

**Comment by cahrens** [comment]
@gpcarr my PR has merged. Let me know if you have questions about some of the changes I made.

**Comment by cahrens** [comment]
@gpcarr You need to rebase this branch on dev and resolve the conflicts. I changed several of the lines you were also editing (we no longer do a separate AJAX call for galaxy disks, and do not want to reintroduce that).

File: src/pages/Environments.js
Comment (gpcarr) [reply src/pages/Environments.js]: This PR is to fix a bug where in workspaces that were created before the project per workspace migration cause a Leo bug. Say a user has disk A in workspace A, deletes workspace A and then creates a new Galaxy in workspace B. Leo tries to attach the same disk from workspace A to the newly created Galaxy in workspace B. In Leo, two disks can't exist in the same workspace. So the fix is that there should be a new disk that gets created for workspace B which post PPW is the case for free

**Comment by gpcarr** [comment]
> @gpcarr You need to rebase this branch on dev and resolve the conflicts. I changed several of the lines you were also editing (we no longer do a separate AJAX call for galaxy disks, and do not want to reintroduce that).

I rebased yesterday! Working on fixing some code before I push that up since a lot of logic around disks was changed while I had to put this PR on pause.

File: src/pages/Environments.js
Comment (cahrens) [reply src/pages/Environments.js]: Thanks @gpcarr, but I am still a bit confused about the issue. Looking at IA-2812, it seems that the issue is that 2 disks cannot exist in the same Google project. With pre-PPW workspaces, you could have 2 workspaces using the same project.

If that is the bug, why isn't it enough to just use googleProject? 

I am also not clear where the limitation is coming from (and therefore I'm unsure if this is something we need to be concerned about with Cromwell also). Can you clarify what the limitation is in Leo code itself?

Comment (cahrens) [reply src/pages/Environments.js]: I also am not following this:

_"Say a user has disk A in workspace A, deletes workspace A and then creates a new Galaxy in workspace B. Leo tries to attach the same disk from workspace A to the newly created Galaxy in workspace B. In Leo, two disks can't exist in the same workspace. So the fix is that there should be a new disk that gets created for workspace B which post PPW is the case for free"_

This seems to be a case of a single disk being used in 2 workspaces (one of which has been deleted). Why is this a problem? How is it related to the Leo limitation that 2 disks can't exist in the same workspace (we have only 1 disk in play)?

Comment (gpcarr) [reply src/pages/Environments.js]: This is a Leo specific limitation. See [this line](https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala#L167) in Leo - we do not allow a disk to exist in two different workspaces. which is currently possible to do with pre-ppw migration workspaces. so we need to force the frontend to create a new disk if a user has 2 workspaces in the same project (these would be older ie pre-ppw workspaces)

File: src/libs/runtime-utils.js
Comment (cahrens) [inline src/libs/runtime-utils.js:287]: Please add unit test coverage for this argument.

File: src/pages/Environments.js
Comment (cahrens) [inline src/pages/Environments.js]: If you use `getDiskAppType`, it takes care of the case sensitivity issue for you. So this would become 
`_.filter(disk => getDiskAppType(disk) === tools.galaxy.appType …`

File: src/pages/Environments.js
Comment (cahrens) [inline src/pages/Environments.js]: appType here should be the upper-case version ("GALAXY"), so you do not want to convert tools.galaxy.appType to lower case.

**Comment by gpcarr** [comment]
> With this PR's changes, I see that I can now create two `Galaxy`s in the same project, using distinct workspaces. I've noticed two issues:
> 
> 1. The Cloud Environments page still displays the warning triangles for having multiple cloud envs/disks.
> 
> image
> 
> 2. When I had the first Galaxy app running and was going to create the second Galaxy app, the Galaxy modal seemed to have selected the existing disk as shown below although a new disk ended up being created. Perhaps there's still something to sort out with the logic that detects an existing disk?
> 
> image
> 
> For testing, I used Rob's pre-PPW workspaces, `test-rstudio` and `rstudio test`, and the `callisto-dev` project.

Hi @kyuksel,

I've addressed all your feedback at this point. The only thing I was not able to recreate is what you have listed as #2 in this comment. Let me know if/how you're still able to!

File: src/pages/Environments.js
Comment (gpcarr) [reply src/pages/Environments.js]: thanks for the feedback, will update!

File: src/libs/runtime-utils.js
Comment (gpcarr) [reply src/libs/runtime-utils.js:287]: will do

**Comment by kyuksel** [comment]
@gpcarr could you update the PR description?

**Comment by kyuksel** [comment]
When I have Galaxy apps `Provisioning` (2 in `pre-PPW` and 1 in `post-PPW` workspaces), I see the warning triangle and the tooltip, which I'd think shouldn't have happened?

image

**Comment by gpcarr** [comment]
> @gpcarr could you update the PR description?

just updated it

File: src/pages/Environments.js
Comment (kyuksel) [inline src/pages/Environments.js]: Looks like we had already gotten the app type. Can we use that instead of getting it again?
```suggestion
              const galaxyDisksOnly = _.filter(disk => appType === tools.galaxy.appType && disk.status !== 'DELETING', filteredDisks)
```

File: src/pages/Environments.js
Comment (kyuksel) [inline src/pages/Environments.js:462]: Within the `cellRenderer` function, while evaluating whether we have multiple disks, we seem to be running both of the evaluation logic (one for Galaxy apps and one for non-Galaxy apps). This should be unnecessary since an app cannot be Galaxy and non-Galaxy at the same time. In other words we could move the conditional from the rendering code to higher up and avoid the unnecessary computations, like so:

```javascript
            cellRenderer: ({ rowIndex }) => {
              const { status: diskStatus, googleProject } = filteredDisks[rowIndex]
              const appType = getDiskAppType(filteredDisks[rowIndex])
              const haveMultipleDisks = () => {
                if (appType === tools.galaxy.appType) {
                  // Galaxy can have multiple disks per project as long as they are in different workspaces
                  const galaxyDisksOnly = _.filter(disk => appType === tools.galaxy.appType && disk.status !== 'DELETING', filteredDisks)
                  const galaxyDiskWorkspaces = _.map(currentDisk => currentDisk.labels.saturnWorkspaceName, galaxyDisksOnly)
                  return _.uniq(galaxyDiskWorkspaces).length !== galaxyDiskWorkspaces.length
                } else {
                  return _.remove(disk => getDiskAppType(disk) !== appType || disk.status === 'Deleting', disksByProject[googleProject]).length > 1
                }
              }

              const forAppText = !!appType ? ` for ${_.capitalize(appType)}` : ''

              return h(Fragment, [
                googleProject,
                diskStatus !== 'Deleting' && haveMultipleDisks() &&
                h(TooltipTrigger, {
                  content: `This billing project has multiple active persistent disks${forAppText}. Only the latest one will be used.`
                }, [icon('warning-standard', { style: { marginLeft: '0.25rem', color: colors.warning() } })])
              ])
            }
```

What do you think?

Comment (kyuksel) [reply src/pages/Environments.js:462]: I have made the PR #2750 into your branch to better illustrate the change above and also pull the logic into its own function, which would also make it possible to write a unit test for it.

File: src/pages/Environments.js
Comment (cahrens) [reply src/pages/Environments.js]: @gpcarr @kyuksel the Leo line of code linked above is not specific to Galaxy-- it is used by both Cromwell and Galaxy (see line 154). Therefore I think these changes should be made general for all apps, not Galaxy-specific. I have some questions about what a Custom app instance would do, but I don't think that is exposed to users, and I doubt the persistent disk logic is implemented for Custom. For simplicity, I think it would be best to make the restriction consistent for all apps until we have a use case where it would not be wanted.

File: src/components/GalaxyModal.js
Comment (cahrens) [inline src/components/GalaxyModal.js:49]: To make Cromwell consistent with Galaxy, it will be necessary to also make this change in CromwellModal.

File: src/libs/_tests/runtime-utils.test.js
Comment (cahrens) [inline src/libs/_tests/runtime-utils.test.js:297]: I believe we will want to make the workspace name a required variable for getCurrentPersistentDisk (update all test usages as well).

File: src/pages/Environments.js
Comment (cahrens) [reply src/pages/Environments.js]: I think this check should not be Galaxy-specific.

File: src/pages/Environments.js
Comment (cahrens) [reply src/pages/Environments.js]: (I admit though that I still don't understand **why** Leo is enforcing this restriction)

**Comment by kyuksel** [comment]
@gpcarr Posting here for you, I and @cahrens to get on the same page regarding the next steps about Christina's [PR comment](https://github.com/DataBiosphere/terra-ui/pull/2662/files#r778847503) on removing app specificity. You're welcome to take your time to address it but we're also available for mobbing, etc. if you feel it'd help at some point 😊

File: src/libs/_tests/runtime-utils.test.js
Comment (cahrens) [inline src/libs/_tests/runtime-utils.test.js]: Please make sure the dates are unique and don't impact the existing tests. The dates of disks are used in some of our logic.

File: src/libs/_tests/runtime-utils.test.js
Comment (cahrens) [inline src/libs/_tests/runtime-utils.test.js]: In general, `saturnApplication` should be "GALAXY". I had one with "galaxy" to represent a certain condition (which is why I had the comment after it), but we don't want to litter that all of the place because it makes it more confusing.

File: src/libs/_tests/runtime-utils.test.js
Comment (cahrens) [inline src/libs/_tests/runtime-utils.test.js]: I think the mocks should be realistic, so please make sure that names are unique.

File: src/libs/runtime-utils.js
Comment (cahrens) [inline src/libs/runtime-utils.js]: I would add documentation to this method to clarify that it does not check the app types, and therefore it is expected that the disks passed in have already been filtered by app type. And I would suggest making the method name `workspaceHasMultipleDisks`.

File: src/pages/Environments.js
Comment (cahrens) [inline src/pages/Environments.js]: Is inactive still needed given line 206?

File: src/libs/runtime-utils.js
Comment (cahrens) [inline src/libs/runtime-utils.js]: Same comment about documentation and name as above.

File: src/libs/_tests/runtime-utils.test.js
Comment (cahrens) [inline src/libs/_tests/runtime-utils.test.js]: Note this is confusing because mockAppDisks contains disks related to different app types.

I really would be tempted to have `workspaceHasMultipleDisks` do the app filtering itself (so you would pass in the appType as a variable) because I think it makes the method much easier to understand.

File: src/libs/runtime-utils.js
Comment (cahrens) [reply src/libs/runtime-utils.js]: Please add unit test coverage for this as well.

Also, the implementation is the same for the 2 methods, so you could consider refactoring them to share the implementation, as long as you can make the contract of what is passed in clear.

(Note though that I'm somewhat in favor of making the methods do the appType filtering, which if you go down that route could make sharing the code more complicated.)

File: src/libs/_tests/runtime-utils.test.js
Comment (gpcarr) [reply src/libs/_tests/runtime-utils.test.js]: I like the idea of changing the method to do the app filtering, i'll update!

**Comment by kyuksel** [comment]
Warning triangles seem to still appear when I create Galaxy apps in distinct workspaces that belong to the same billing project?

image

**Comment by gpcarr** [comment]
> Warning triangles seem to still appear when I create Galaxy apps in distinct workspaces that belong to the same billing project?
> 
> image

I'm not seeing that on my end. Are you on this branch with the latest changes?

**Comment by kyuksel** [comment]
> I'm not seeing that on my end. Are you on this branch with the latest changes?

Yes. I think the warnings were due to the disks in `Deleting` status left over from error'ed apps. The warnings went away once I marked the disks `Deleted` in Leo DB. This shouldn't have been happening though, right, since we're supposed to exclude `Deleting` disks from consideration? I wonder if we can replicate the issue in a unit test.

File: src/pages/Environments.js
Comment (kyuksel) [inline src/pages/Environments.js]: We use `!!` to be explicit that we're coercing a non-boolean (`appType`) onto a boolean.

```suggestion
    return !!appType ? workspaceHasMultipleDisks(disks, appType) : _.remove(disk => getDiskAppType(disk) !== appType || disk.status === 'Deleting',
```

Comment (kyuksel) [reply src/pages/Environments.js]: Would `getDiskAppType(disk) !== appType` ever evaluate to `true` since it's in the second branch of the ternary where `!!appType` would be `false`?

Comment (cahrens) [reply src/pages/Environments.js]: I think the code is iterating over all the disks, and we want to remove the ones that aren't associated with this specific appType.

Comment (kyuksel) [reply src/pages/Environments.js]: Right. I was trying to understand, for the case where `!!appType` is `false` (meaning, `appType` is `undefined`?), how `getDiskAppType(disk) !== appType` could be `true`. Looking more closely, it looks like `getDiskAppType(disk)` returns `undefined` if `disk` is unattached. In other words, `getDiskAppType(disk) !== appType` is necessary to filter out unattached disks?

Comment (cahrens) [reply src/pages/Environments.js]: Not just unattached disks, but disks related to a different app. Like Galaxy vs. Cromwell.

Comment (kyuksel) [reply src/pages/Environments.js]: `_.remove(disk => getDiskAppType(disk) !== appType || disk.status === 'Deleting'…` part of the ternary would not be executed if a disk has `appType === Galaxy/Cromwell`.

Comment (gpcarr) [reply src/pages/Environments.js]: plan on discussing this comment on a call at noon today with @kyuksel if you'd like to join @cahrens

Comment (kyuksel) [reply src/pages/Environments.js]: For posterity: Talked to Gabriela offline and sorted out my misunderstanding.

**Comment by gpcarr** [comment]
@kyuksel @cahrens all feedback has been addressed, looking for a final review/thumb :) thanks!

File: src/pages/Environments.js
Comment (cahrens) [inline src/pages/Environments.js]: I think the argument can be just `appType` instead of `app.appType`.

File: src/pages/Environments.js
Comment (cahrens) [inline src/pages/Environments.js]: I believe `forAppText` is defined elsewhere now.

Comment (gpcarr) [reply src/pages/Environments.js]: just updated this!

File: src/pages/Environments.js
Comment (cahrens) [inline src/pages/Environments.js:326]: I think a comment would be good to explain that we expect appType to be `undefined` in the case of runtimes (and therefore the second part of the conditional is filtering out disks that are app disks instead of runtime disks).

Comment (gpcarr) [reply src/pages/Environments.js:326]: i'll add one

**Review by cahrens — APPROVED** [review]
The code LGTM. I'm not able to test the original issue (multiple disks in the same workspace), so I'm leaving that to others.

I love how the code got simpler in each iteration! And thanks for the tests.

**Review by kyuksel — APPROVED** [review]
Thanks for addressing our comments 😊

**Pull request merged**
